src/binary_heap.rs

```diff
@@ -1086,6 +1097,132 @@ impl<T, C: Compare<T>> BinaryHeap<T, C> {
 
         self.rebuild_tail(start);
     }
+
+    /// Retains only the elements specified by the predicate.
+    ///
+    /// In other words, remove all elements `e` for which `f(&e)` returns
+    /// `false`. The elements are visited in unsorted (and unspecified) order.
+    ///
+    /// # Examples
+    ///
+    /// Basic usage:
+    ///
+    /// ```
+    /// use std::collections::BinaryHeap;
+    ///
+    /// let mut heap = BinaryHeap::from([-10, -5, 1, 2, 4, 13]);
+    ///
+    /// heap.retain(|x| x % 2 == 0); // only keep even numbers
+    ///
+    /// assert_eq!(heap.into_sorted_vec(), [-10, 2, 4])
+    /// ```
+    // #[stable(feature = "binary_heap_retain", since = "1.70.0")]
+    pub fn retain<F>(&mut self, mut f: F)
+    where
+        F: FnMut(&T) -> bool,
+    {
+        // rebuild_start will be updated to the first touched element below, and the rebuild will
+        // only be done for the tail.
+        let mut guard = RebuildOnDrop {
+            rebuild_from: self.len(),
+            heap: self,
+        };
+        let mut i = 0;
+
+        guard.heap.data.retain(|e| {
+            let keep = f(e);
+            if !keep && i < guard.rebuild_from {
+                guard.rebuild_from = i;
+            }
+            i += 1;
+            keep
+        });
+    }
+
+    /// Removes the first element specified by the predicate.
+    ///
+    /// In other words, remove the first element `e` for which `f(&e)` returns
+    /// `true`. The elements are visited in unsorted (and unspecified) order.
+    ///
+    /// # Examples
+    ///
+    /// Basic usage:
+    ///
+    /// ```
+    /// use binary_heap_plus::BinaryHeap;
+    ///
+    /// let mut heap = BinaryHeap::from([-10, -5, 1, 2, 4, 13]);
+    ///
+    /// heap.remove_once(|x| *x == -5); // remove -5
+    ///
+    /// assert_eq!(heap.into_sorted_vec(), [-10, 1, 2, 4, 13])
+    /// ```
+    pub fn remove_once<F>(&mut self, mut f: F)
+    where
+        F: FnMut(&T) -> bool,
+    {
+        let mut guard = RebuildOnDrop {
+            rebuild_from: self.len(),
+            heap: self,
+        };
+        let mut i = 0;
+
+        for e in &guard.heap.data {
+            let remove = f(e);
+
+            if remove && i < guard.rebuild_from {
```

Comment:
Also applies to replace_once

```suggestion
            if remove {
```